### ISSUE TITLE
Implement common optimizers

### ISSUE THREAD

**Issue by bwasti** [facebookresearch/shumai#19]
Similar to the module task, there are many commonly used optimizers that shouldn't be rewritten by users

In keeping with the JS ways of doing things, two features should be provided
1. Stateful "parameter list" based optimization (for object oriented code)
2. Completely automatic deduced parameters based on the gradient graph (for functional programming)

Here's a functional version of optimize:
```javascript
let c = new CustomModule(...)
let l = new Linear(128, 128)
function model(x) {
  x = c(x)
  x = l(x)
  return sm.relu(x)
}

for (let i = 0; i < iters; ++i) {
  const loss = mse(model(example[i]), ref[I])
  zero_grad(loss) // maybe make this optional too?
  loss.backward()
  optimize(loss, {
    exclude: c.exclude_me_sometimes
  })
}
```
